src/bin/genemichaels.rs

```diff
@@ -250,14 +330,147 @@ fn main() {
             };
         }
+        eprintln!(
+            "\x1B[1;32m    Finished\x1B[0;22m workspace formatted successfully in {:.2}s",
+            time::Instant::now().duration_since(inst).as_secs_f64()
+        );
         if failed {
             process::exit(1);
         }
     }
 }
+
+fn process_cargo_toml(
+    path: PathBuf,
+    manifest: Manifest,
+    thread_count: Option<usize>,
+    config: FormatConfig,
+) -> Result<()> {
+    let mut dirs = Vec::from([]);
+    for bin in manifest.bin.into_iter().flatten() {
+        if let Some(bin_path) = bin.path {
+            let bin_path =
+                bin_path
+                    .parse::<PathBuf>()?
+                    .parent()
+                    .map(|pp| pp.to_path_buf())
+                    .ok_or_else(|| anyhow::anyhow!("Unable to get parent of {bin_path}"))?;
+            dirs.push(bin_path);
+        }
+    }
+    if let Some(lib) = manifest.lib {
+        if let Some(p) =
+            lib
+                .path
+                .and_then(|p| p.parse::<PathBuf>().ok())
+                .and_then(|p| p.parent().map(|pp| pp.to_path_buf())) {
+            dirs.push(p)
+        }
+    }
+    for bench in manifest.bench.into_iter().flatten() {
+        if let Some(bench_path) = bench.path {
+            let bench_path =
+                bench_path
+                    .parse::<PathBuf>()?
+                    .parent()
+                    .map(|pp| pp.to_path_buf())
+                    .ok_or_else(|| anyhow::anyhow!("Unable to get parent of {bench_path}"))?;
+            dirs.push(bench_path);
+        }
+    }
+    for test in manifest.test.into_iter().flatten() {
+        if let Some(test_path) = test.path {
+            let test_path =
+                test_path
+                    .parse::<PathBuf>()?
+                    .parent()
+                    .map(|pp| pp.to_path_buf())
+                    .ok_or_else(|| anyhow::anyhow!("Unable to get parent of {test_path}"))?;
+            dirs.push(test_path);
+        }
+    }
+    for example in manifest.example.into_iter().flatten() {
+        if let Some(example_path) = example.path {
+            let example_path =
+                example_path
+                    .parse::<PathBuf>()?
+                    .parent()
+                    .map(|pp| pp.to_path_buf())
+                    .ok_or_else(|| anyhow::anyhow!("Unable to get parent of {example_path}"))?;
+            dirs.push(example_path);
+        }
+    }
+    if let Some(ws) = manifest.workspace {
+        let workspace_dirs: Vec<PathBuf> =
+            ws.members.into_iter().filter_map(|m| path.join(&m).exists().then(|| path.join(m))).collect();
+
+        // loop through each folder in the workspace and recursively run the formatter
+        for workspace in workspace_dirs {
+            let manifest = cargo_manifest::Manifest::from_path(workspace.join(CARGO_TOML))?;
+            process_cargo_toml(workspace, manifest, thread_count, config)?
```

Comment:
This works, but it means that all of one manifest's files need to finish formatting before moving on to the next manifest.

I'd do the `dirs` aggregation in a recursive function but do the parallel formatting on the whole dir list after the recursion is done - maybe the function can return the dirs array for the manifest it's given.

Reply:
Sorry, keep thinking of new things.  One other option is, rather than add directories to the vector as you go, you could add spawned jobs to the pool (that is, pass a mutable pool + seen dir set around, and spawn the jobs immediately when you encounter the directory).

Then, at the end, just join the pool.

Reply:
I think the first option is clearer (and easier to write) so that's what I've implemented (using HashSet instead of Vec).

It seems to work as expected. Although I didn't see any real performance difference, I agree it makes sense to run all the jobs at the same time for maximum efficiency.

Reply:
Ah right, need the hashset for dedup anyways, that's a good idea.  Great!